blog/2020/09/rr-memory-magic.md

```diff
@@ -0,0 +1,503 @@
+
+If the replay fails on the original machine as well (it did), this is some evidence against a microarchitectural
+issue (it's not guaranteed of course, but executing the exact same instruction sequence at least has
+some chance to re-trigger such an issue). As for the second recording, it failed in exactly the same
+way: A bit flip in the low bit of the high byte of some pointer. This strongly implicates faulty memory
+as the fault is in the same bit position across both runs, which could happen, if one particular bit of
+a memory cell is degraded (e.g. in such a way that charge leaks into it over time, or there is some
+damage to the particular data line that this bit is on). In particular, if the issue were radiation-induced
+(perhaps the user's office happens to be above the physics department fusion reactor), we would have
+not expected a preference for any particular bit index.
+
+Having excluded these other issues to my satisfaction, I closed the report with my conclusion and
+encouraged the user to run memtest on their machine. And indeed, the next day I got the following
+screenshot:
+
+~~~
+<img id="meme" src="/assets/blog/2020-09-23-rr-memory/faulty-mem.jpg"/>
+~~~
+
+The failures reported are exactly what we expected: A 0->1 bit flip in the low bit of the high
+byte (second-highest nibble) of the memory location. Since all the failing memory addresses have
+then `0x5c` address bits set, we can probably make the assumption that the bit that was flipped
+was that of the array itself (whose address ended in `0x8`, i.e. having the high dword at `0xc`),
+rather than the copy of the type tag in the svec (whose address ended in `0x0`), but none of the
+detected failures exactly match the address offset of one of the two addresses we identified as
+potential targets for the bit flip (though the test in the screenshot was not run to completion,
+so maybe it would have found one). Further analysis of the failure would require physical access
+to the DRAM and perhaps an electron microscope, but we need not go there, because we've already
+determined this is not a Julia bug and the user will probably be replacing their machine.
+
+# Conclusion
+
+With about an hour's work (plus 2-3 hours to add the `seek-to-ticks` functionality written,
+tested, upstreamed, etc.), we managed to remotely and conclusively diagnose that a reported
+crash was caused by faulty memory on the users machine.  Even though the rr trace did not
+contain the actual cause of the crash, it was nevertheless
+an invaluable tool to analyze this issue. The fact that the stray bit flip was missing
+from the recording already excluded 99% of possible cases for the crash, and some
+careful analysis of the trace gave enough clues to be able to eliminate most of the
+remaining. To quote Sherlock Holmes "When you have eliminated the impossible,
+whatever remains, however improbable, must be the truth". Without rr, it may often
```

Comment:
```suggestion
remaining. To quote Sherlock Holmes
> When you have eliminated the impossible, whatever remains, however improbable, must be the truth.
Without `rr`, it may often
```